symengine/functions.cpp

```diff
@@ -329,6 +329,95 @@ bool inverse_lookup(umap_basic_basic &d, const RCP<const Basic> &t,
     }
 }
 
+Sign::Sign(const RCP<const Basic> &arg) : OneArgFunction(arg)
+{
+    SYMENGINE_ASSIGN_TYPEID()
+    SYMENGINE_ASSERT(is_canonical(arg))
+}
+
+bool Sign::is_canonical(const RCP<const Basic> &arg) const
+{
+    if (is_a_Number(*arg)) {
+        return false;
+    }
+    if (is_a<Sign>(*arg)) {
+        return false;
+    }
+    if (is_a<Mul>(*arg)) {
+        return false;
+    }
+    if (is_a<Pow>(*arg)
+        and is_a_Complex(*down_cast<const Pow &>(*arg).get_base())) {
+        return false;
+    }
+    return true;
+}
+
+RCP<const Basic> Sign::create(const RCP<const Basic> &arg) const
+{
+    return sign(arg);
+}
+
+RCP<const Basic> sign(const RCP<const Basic> &arg)
+{
+    if (is_a_Number(*arg)) {
+        if (is_a<NaN>(*arg)) {
+            return Nan;
+        }
+        if (down_cast<const Number &>(*arg).is_zero()) {
+            return zero;
+        }
+        if (down_cast<const Number &>(*arg).is_positive()) {
+            return one;
+        }
+        if (down_cast<const Number &>(*arg).is_negative()) {
+            return minus_one;
+        }
+        if (is_a_Complex(*arg)
+            and down_cast<const ComplexBase &>(*arg).is_re_zero()) {
+            RCP<const Basic> r = mul(mul(minus_one, I), arg);
+            if (is_a_Number(*r)
+                and down_cast<const Number &>(*r).is_positive()) {
+                return I;
+            }
+            if (is_a_Number(*r)
+                and down_cast<const Number &>(*r).is_negative()) {
+                return mul(minus_one, I);
+            }
+        }
+    }
+    if (is_a<Sign>(*arg)) {
+        return arg;
+    }
+    if (is_a<Mul>(*arg)) {
+        const Mul &r = down_cast<const Mul &>(*arg);
+        RCP<const Basic> s = sign(r.get_coef());
+        map_basic_basic dict = r.get_dict();
+        RCP<const Number> coef = one;
+        map_basic_basic new_dict;
+        for (const auto &p : dict) {
+            if (is_a_Complex(*p.first)
+                and down_cast<const ComplexBase &>(*arg).is_re_zero()) {
```

Comment:
The tests pass locally on changing the line to 
`RCP<const Basic> r = down_cast<const ComplexBase &>(*arg).imaginary_part();`
Thanks @ranjithkumar007!